### ISSUE TITLE
Optimize first meaningful paint

### ISSUE THREAD

**Issue by gonzalezjo** [scs-ben/tcec-gui#10]
[First meaningful paint (FMP)](https://developers.google.com/web/fundamentals/performance/user-centric-performance-metrics#first_meaningful_paint_and_hero_element_timing) describes the time delta between when one clicks on a page, and when the most meaningful element(s) on the page, or "hero elements", are rendering and useful. 

For the TCEC GUI, there are a few hero elements. The most obvious one is **the chessboard**. However, **the schedule** is also important, especially for fans of a particular engine, which might want to see when and if their favorite is playing, before leaving the page. Then, **the scoreboard** would be the next hero element, for users that want to quickly check in on how their favorite engines are performing. 

There are two approaches to optimizing first meaningful paint: one, by micro-optimizing the loading of specific hero elements, or two, by seeking more general performance gains. (e.g., using compression and CDNs.) 

Based on my analysis, the latter will reap the most rewards. However, there are still notable gains to be had from a combination of approaches. For analysis, scroll down to the reply to this issue.

**Comment by gonzalezjo** [comment]
## Analysis
### General performance concerns

1. No CDN usage 


This isn't a request to add Cloudflare or similar to the entire site. While that would probably help bring load times down a good degree, especially for time-to-first-byte, (establishing a connection quickly is very helpful for users on poor 3G connections) simply loading JavaScript from cdnjs, unpkg, MaxCDN, bootstrapcdn, etc., would be great. This is two-fold: one, they have peering with ISPs, meaning great download speeds, and two, since lots of websites use these JavaScript CDNs, it increases our chance at a cache hit. 


It is important that we do not use too many different CDNs, because TLS negotiation can take time, especially on 2G/3G. If you only need to do it once, because you're loading all of your JS from one server, that's optimal.


2. No HTTP2


HTTP2 should provide substantial performance improvements, thanks to multiplexing and the amount of unminified JS being served.



3. Cache usage


Smart caching is a massive boost to site performance. Use of caching can make repeated site visits *much* faster. [Smart use of caching can do even more](https://code.fb.com/web/this-browser-tweak-saved-60-of-requests-to-facebook/). Essentially, most assets will be kept static for years. For example, the chess pieces and most of the engine logos are not going to change. On the off chance that they will change, we would still like to serve updated versions to users. The Facebook approach to this was to hash these asset names, and request them by their hash. We don't have to do that, because we're not at Facebook's scale. We can simply name images something like, "arasan-october-7-2018.png," and update the file name if we ever want to get the browser not to load a stale cache entry.



4. Minification

Minifying JS would help reduce latency of JS loading by a decent amount. It would also decrease parse time, which seems to take about 110 ms on my i7-4790K. Presumably, it's worse on mobile.


5. Using less JavaScript


I counted about 25 different, non-Twitch scripts being loaded. Twitch adds a few more, but they're loaded in an iframe, which shouldn't be blocking anything but onload. It blocks page load for many seconds on my phone, even on 4G. Surely, not all of these are needed?



6. 

### Specific performance concerns


1. Compression


Though this is also a more general performance improvement, it's a specifically nice boost for updatePgn(), so I'm including it here. Enabling Brotli and gzip would help quite a bit with updatePgn.



2. Investigate Twitch 

There are two things here:


1. I've found this hard to debug because I can't tell what it's related to, but I'm pretty sure that the loading of the Twitch Chat frame delays the loading of Google Ads and other elements, because of a quirk with [window.onload](http://www.lognormal.com/blog/2012/12/12/the-script-loader-pattern/). This leads to some notable jank. It might be something else, though, but it's really hard to tell, because it only happens on my phone with poor signal.


2. Loading the actual Twitch video player might just suck up bandwidth and delay page load for no reason. It takes about 110 ms on an i7-4790K, I think.


3. Asset loading


The page is blocked on load by stylesheets and some other assets. Some blocking events can be made asynchronous or deferred. Non-asynchronous and non-deferred scripts make up 450 ms of the time spent attempting to finish parsing the HTML on my i7-4790K, so again, that would be worse on a phone. Some thought was given to this, apparently, based on the "Placed at the end of the document so the pages load faster" comment in live.html. 



4. Chessboard micro-optimization


Sending a FEN to draw before sending the rest of the data would make the page load seem perceptually faster.



5. Send less JSON.


In endgames, the JSON payload for livedata can easily reach over a megabyte. In Chiron v. Komodo, game 26, which I am watching now, it requests and parses 1.1 megabytes of JavaScript every 3 seconds, *just* for livedata.json. This sounds bad because of bandwidth, but it's also pretty bad for performance. 


According to Chromium's profiler, it leads to frequent major garbage collections



6. Optimize chart rendering


Calls into chart.min.js can be extremely intensive. On an i7, I regularly see it taking 50 milliseconds. An anonymous function calls startDigest, which calls `advance` and `requestAnimationFrame`. `advance` takes a *lot* of time. This can lead to jank during page load.

Unrelated to FMP time, we seem to call into chart.min.js with a very high frequency, even when nothing is happening. I have no idea why or how this is happening, and I can't find where in the source that this takes place, but it is. In a 30000 ms period, 890.4 ms of CPU time was spent on drawing chart animations, even when nothing was happening visually. This was on an i7 & GTX 1050, which Chromium should be able to use for HW acceleration. This might be really hard on mobile devices. Should I be making a separate issue for this?

**Comment by gonzalezjo** [comment]
## Standardized Analysis


Performance metrics for analyzing page load speeds exist. I ran a battery of them against the site for two reasons. One, to see if it would pick up on anything, and two, for evaluation of future results.


[Uptrends](https://www.uptrends.com/tools/website-speed-test): [Click here](https://user-images.githubusercontent.com/22648562/45258309-64ead780-b383-11e8-8a34-a56ff33cab46.png)

Summary: PageSpeed score of 21. Ouch.



[Pingdom](https://tools.pingdom.com): [Click here](https://tools.pingdom.com/#!/dE7urD/http://tcec.chessdom.com/beta/live.html)

Summary: Their bot had to load 24.2 MB of resources before it had finished. It made 53 requests for scripts.


[Google PageSpeed Insights](https://developers.google.com/speed/pagespeed/insights/): [Click Here](https://developers.google.com/speed/pagespeed/insights/?url=http%3A%2F%2Ftcec.chessdom.com%2Fbeta%2Flive.html) 

Summary: No bueno. There were four occurrences in which page rendering was completely blocked while waiting on a network resource. 



[GTMetrix](https://gtmetrix.com/): [Click here](https://gtmetrix.com/reports/tcec.chessdom.com/tC2xH9RY) 


Summary: We got given an F. [Autogenerated report](https://github.com/scs-ben/tcec-gui/files/2363698/GTmetrix-report-tcec.chessdom.com-20180908T131934-tC2xH9RY-full.pdf)



[Google Test My Site](https://testmysite.thinkwithgoogle.com/intl/en-us): Report pending.






### Commentary


It seems like we fetch a lot of stuff, we do so from a lot of origins, we don't compress or cache much of that stuff effectively, and do so in the way of rendering important assets. 


The good news is, much of that is easily fixed. Compression and caching, for example, are basically just headers. Minification is simple. 


The bad news is, some of those aren't easily fixed, and they have a huge impact on the site's speed.


Uptrends generated this interesting table: 



```
Total bytes: 12201.5 kB
Requests: 199
Resolve
2009 ms / 199 requests ~ 10.1 ms (average)
TCP Connect
1716 ms / 199 requests ~ 8.6 ms (average)
HTTPS Handshake
2339 ms / 199 requests ~ 11.8 ms (average)
Send
11 ms / 199 requests ~ 0.1 ms (average)
Wait
13979 ms / 199 requests ~ 70.2 ms (average)
Receive
24502 ms / 199 requests ~ 123.1 ms (average)
````


It does a good job summing up the issue. We make hundreds of requests, just loading the page for the first time. Reducing that number is going to be very difficult and tedious. We spend 14 seconds stalling on these requests, and an additional 10 seconds are spent downloading, a lot of data, which we can't really avoid with the current system. 


Google PageSpeed Insights gave us this, regarding the mobile experience: "Your page has 15 blocking script resources and 7 blocking CSS resources. This causes a delay in rendering your page." That might be difficult to address.




### Commentary

We have a lot of low-hanging fruit, and a lot of not-so-low-hanging-fruit. Automated analysis did a good job at detecting some more egregious issues. Not all of these are easy to fix, but they're a very good place to start, and provide a good way to measure progress. I think this is a good place to start working on PRs. Expect some from me soon. :stuck_out_tongue:

**Comment by scs-ben** [comment]
Game now loads within 2 seconds with cache enabled. I think this is resolved
